File: scripts/run-evals.py
#!/usr/bin/env python3
"""Static golden-task eval harness for Hyperflow repo contracts."""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
TASKS = ROOT / "evals" / "tasks"


def skill_dirs() -> set[str]:
    return {p.parent.name for p in (ROOT / "skills").glob("*/SKILL.md")}


def check(name: str) -> tuple[bool, str]:
    if name.startswith("file:"):
        rel = name.split(":", 1)[1]
        ok = (ROOT / rel).is_file()
        return ok, rel
    if name.startswith("readme_contains:"):
        needle = name.split(":", 1)[1]
        text = (ROOT / "README.md").read_text(encoding="utf-8")
        ok = needle in text
        return ok, f"README contains {needle!r}"
    if name.startswith("readme_not_contains:"):
        # unused simple path; keep for extension
        needle = name.split(":", 1)[1]
        text = (ROOT / "README.md").read_text(encoding="utf-8")
        ok = needle not in text
        return ok, f"README lacks {needle!r}"
    if name == "skill_count_readme":
        actual = len(skill_dirs())
        readme = (ROOT / "README.md").read_text(encoding="utf-8")
        # e.g. Nineteen skills · 19
        m = re.search(r"(\d+)\s*\)?\s*$", "")  # placeholder
        m = re.search(r"skills[^\n]*·\s*(\d+)", readme) or re.search(
            r"\((\d+)\)\s*$", ""
        )
        m = re.search(r"## Skills \([^·]*·\s*(\d+)\)", readme)
        if not m:
            m = re.search(r"·\s*(\d+)\s*\)", readme)
        if not m:
            return False, "could not parse skill count from README"
        n = int(m.group(1))
        return n == actual, f"README {n} vs dirs {actual}"
    if name == "features_skills_match_dirs":
        data = json.loads((ROOT / "config" / "features.json").read_text(encoding="utf-8"))
        registered = set()
        for entry in data.get("skills", []):
            if isinstance(entry, str):
                registered.add(entry)
            elif isinstance(entry, dict) and "name" in entry:
                registered.add(entry["name"])
            elif isinstance(entry, dict) and "id" in entry:
                registered.add(entry["id"])
        dirs = skill_dirs()
        if registered != dirs:
            return False, f"only_in_features={sorted(registered-dirs)} only_in_dirs={sorted(dirs-registered)}"
        return True, f"{len(dirs)} skills"
    return False, f"unknown check {name}"


def load_tasks() -> list[dict]:
    tasks = []
    for path in sorted(TASKS.glob("*.json")):
        tasks.append(json.loads(path.read_text(encoding="utf-8")))
    return tasks


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--list", action="store_true")
    args = ap.parse_args()
    tasks = load_tasks()
    if args.list:
        for t in tasks:
            print(f"{t['id']}: {t.get('title','')}")
        return 0
    failed = 0
    for t in tasks:
        tid = t["id"]
        results = []
        ok_all = True
        for c in t.get("checks", []):
            ok, detail = check(c)
            results.append((ok, c, detail))
            ok_all = ok_all and ok
        status = "PASS" if ok_all else "FAIL"
        if not ok_all:
            failed += 1
        print(f"{status}  {tid}")
        for ok, c, detail in results:
            mark = "ok" if ok else "FAIL"
            print(f"    [{mark}] {c} — {detail}")
    print(f"---\n{len(tasks) - failed}/{len(tasks)} passed")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/run-evals.mjs
#!/usr/bin/env node
/** Static golden-task evaluations for portable Hyperflow contracts. */
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { dirname, extname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const TASKS = join(ROOT, "evals", "tasks");

function repoPath(path) {
  const target = resolve(ROOT, path);
  if (target !== ROOT && !target.startsWith(`${ROOT}/`)) throw new Error(`task path escapes repository: ${path}`);
  return target;
}

function read(path) {
  return readFileSync(repoPath(path), "utf8");
}

function json(path) {
  return JSON.parse(read(path));
}

function markdownTable(path) {
  const fields = new Map();
  for (const line of read(path).split(/\r?\n/)) {
    const match = line.match(/^\|\s*([^|]+?)\s*\|\s*([^|]+?)\s*\|$/);
    if (!match || /^-+$/.test(match[1].trim())) continue;
    fields.set(match[1].trim(), match[2].trim());
  }
  return fields;
}

function gitRefResolves(ref) {
  try {
    execFileSync("git", ["rev-parse", "--verify", `${ref}^{commit}`], { cwd: ROOT, stdio: "pipe" });
    return true;
  } catch {
    return false;
  }
}

function reviewMemory(path) {
  const source = read(path).trim();
  const entries = source.split(/^## /m).slice(1).map((entry) => `## ${entry}`);
  const fields = ["Verdict", "Target", "Audit", "Decision", "Follow-up"];
  const validEntries = entries.length > 0 && entries.length <= 20 && entries.every((entry) => {
    const lines = entry.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
    const fieldLines = lines.slice(1);
    const exactFields = fieldLines.length === fields.length
      && fields.every((field, index) => fieldLines[index].startsWith(`- ${field}: `) && fieldLines[index].length > field.length + 4);
    return lines.length === fields.length + 1
      && /^## \d{4}-\d{2}-\d{2} — .+$/.test(lines[0])
      && /^- Verdict: PASS$/.test(fieldLines[0] ?? "")
      && /^- Target: `[^`]+`$/.test(fieldLines[1] ?? "")
      && /^- Audit: `\.hyperflow\/audits\/[^`]+\.md`$/.test(fieldLines[2] ?? "")
      && exactFields;
  });
  const ok = /^# Review outcomes$/m.test(source) && validEntries;
  return {
    ok,
    detail: ok
      ? `${entries.length} accepted review entr${entries.length === 1 ? "y" : "ies"} within the bounded pointer format`
      : "review memory must contain only bounded PASS entries with exact target/audit pointers",
  };
}

function filesUnder(path) {
  const absolute = repoPath(path);
  if (!existsSync(absolute)) return [];
  if (!statSync(absolute).isDirectory()) return [path];
  return readdirSync(absolute, { withFileTypes: true }).flatMap((entry) => {
    if (entry.name === ".git" || entry.name === ".hyperflow" || entry.name === "node_modules") return [];
    const child = join(path, entry.name);
    if (entry.isDirectory()) return filesUnder(child);
    return entry.isFile() ? [child] : [];
  });
}

function localMarkdownLinksResolve() {
  const documents = ["README.md", "PRIVACY.md", "RELEASING.md", ...filesUnder("docs").filter((path) => extname(path) === ".md")];
  for (const document of documents) {
    const source = read(document);
    for (const match of source.matchAll(/\[[^\]]+\]\(([^)]+)\)/g)) {
      const target = match[1].split("#", 1)[0];
      if (!target || /^(?:https?:|mailto:)/.test(target)) continue;
      if (!existsSync(resolve(dirname(repoPath(document)), decodeURIComponent(target)))) return `${document} -> ${target}`;
    }
  }
  return null;
}

function check(spec) {
  switch (spec.type) {
    case "files_exist": {
      const missing = spec.paths.filter((path) => !existsSync(repoPath(path)));
      return { ok: missing.length === 0, detail: missing.length ? `missing ${missing.join(", ")}` : `${spec.paths.length} paths present` };
    }
    case "read_contains": {
      const ok = read(spec.path).includes(spec.value);
      return { ok, detail: `${spec.path} ${ok ? "contains" : "lacks"} ${JSON.stringify(spec.value)}` };
    }
    case "read_not_contains": {
      const ok = !read(spec.path).includes(spec.value);
      return { ok, detail: `${spec.path} ${ok ? "does not contain" : "contains"} ${JSON.stringify(spec.value)}` };
    }
    case "skill_set": {
      const actual = readdirSync(repoPath("skills"), { withFileTypes: true })
        .filter((entry) => entry.isDirectory() && existsSync(repoPath(`skills/${entry.name}/SKILL.md`)))
        .map((entry) => entry.name)
        .sort();
      const expected = [...spec.expected].sort();
      const ok = JSON.stringify(actual) === JSON.stringify(expected);
      return { ok, detail: ok ? `${actual.length} public skills` : `expected ${expected.join(", ")}; found ${actual.join(", ")}` };
    }
    case "json_key_absent": {
      const document = json(spec.path);
      const ok = !Object.hasOwn(document, spec.key);
      return { ok, detail: `${spec.path} ${ok ? "does not declare" : "declares"} ${spec.key}` };
    }
    case "docs_links_resolve": {
      const problem = localMarkdownLinksResolve();
      return { ok: problem === null, detail: problem === null ? "current Markdown links resolve" : `broken link ${problem}` };
    }
    case "host_claims": {
      const config = json(spec.path);
      const hosts = config.hosts ?? [];
      const ids = hosts.map((host) => host.id);
      const unique = new Set(ids).size === ids.length;
      const complete = hosts.every((host) => ["id", "claim", "status", "evidence"].every((key) => typeof host[key] === "string"));
      const readme = read("README.md").toLowerCase();
      const codexHonest = hosts.find((host) => host.id === "codex-cli")?.status === "preview_uncertified" && readme.includes("preview");
      const opencodeHonest = hosts.find((host) => host.id === "opencode")?.claim === "compatibility shim" && readme.includes("compatibility shim");
      const antigravityHonest = hosts.find((host) => host.id === "antigravity")?.claim === "compatibility shim" && readme.includes("antigravity");
      const ok = hosts.length >= 4 && unique && complete && codexHonest && opencodeHonest && antigravityHonest;
      return { ok, detail: ok ? `${hosts.length} host claims are explicit` : "host claims are incomplete or overstate support" };
    }
    case "handoff_round_trip": {
      const taskPath = join(spec.path, "task.md");
      const handoffPath = join(spec.path, "handoff.md");
      const task = read(taskPath);
      const fields = markdownTable(handoffPath);
      const requiredFields = [
        "status", "task_pointer", "source_branch", "build_branch", "base_ref", "head_ref",
        "created", "updated", "result", "checks", "commits",
      ];
      const fieldsPresent = requiredFields.every((field) => fields.has(field) && fields.get(field) !== "");
      const taskShape = ["^# .+", "^## Outcome$", "^## Tasks$", "^## Verification$"]
        .every((pattern) => new RegExp(pattern, "m").test(task));
      const pointer = fields.get("task_pointer");
      const pointerExpected = `.hyperflow-handoff/${spec.slug}/task.md`;
      const refsResolve = ["base_ref", "head_ref"].every((field) => gitRefResolves(fields.get(field)));
      const statusOk = fields.get("status") === "reviewed";
      const ok = fieldsPresent && taskShape && pointer === pointerExpected && refsResolve && statusOk;
      return {
        ok,
        detail: ok
          ? "reviewed handoff preserves the task pointer and resolvable base/head refs"
          : "handoff package is missing its Markdown task shape, exact pointer, reviewed status, or resolvable refs",
      };
    }
    case "review_memory":
      return reviewMemory(spec.path);
    default:
      return { ok: false, detail: `unknown check type ${spec.type}` };
  }
}

function loadTasks() {
  const tasks = filesUnder("evals/tasks")
    .filter((path) => extname(path) === ".json")
    .sort()
    .map((path) => json(path));
  const ids = new Set();
  for (const task of tasks) {
    if (typeof task.id !== "string" || ids.has(task.id)) throw new Error(`invalid or duplicate eval id: ${task.id}`);
    if (!Array.isArray(task.checks) || task.checks.length === 0) throw new Error(`eval ${task.id} has no checks`);
    ids.add(task.id);
  }
  return tasks;
}

function main() {
  const args = new Set(process.argv.slice(2));
  const tasks = loadTasks();
  if (args.has("--list")) {
    for (const task of tasks) console.log(`${task.id}: ${task.title}`);
    return 0;
  }

  const results = [];
  let failed = 0;
  for (const task of tasks) {
    const checks = task.checks.map((spec) => {
      try {
        return { ...check(spec), spec };
      } catch (error) {
        return { ok: false, detail: error.message, spec };
      }
    });
    const ok = checks.every((result) => result.ok);
    if (!ok) failed += 1;
    results.push({ id: task.id, title: task.title, ok, checks });
  }

  if (args.has("--json")) {
    console.log(JSON.stringify({ passed: tasks.length - failed, total: tasks.length, results }, null, 2));
  } else {
    for (const result of results) {
      console.log(`${result.ok ? "PASS" : "FAIL"}  ${result.id}`);
      for (const item of result.checks) console.log(`    [${item.ok ? "ok" : "FAIL"}] ${item.spec.type} — ${item.detail}`);
    }
    console.log(`---\n${tasks.length - failed}/${tasks.length} evals passed`);
  }
  return failed === 0 ? 0 : 1;
}

try {
  process.exitCode = main();
} catch (error) {
  console.error(`FAIL eval-harness: ${error.message}`);
  process.exitCode = 1;
}
